### DISCUSSION TITLE
Timing

### DISCUSSION THREAD

**Discussion by MichaelPNolan** [marcel-licence/esp32_basic_synth#65] [Q&A] [answered]
I'm going to investigate the i2s_write because that seems to be where timing and delay that could be used for other communications is spent ... but I don't know. 

There is portMAX_DELAY - and I've been wanting to dive into how it keeps to the sample rate and where I can slip in processing so as not to cause sound artifacts by messing up the timing. 

I have issues with clipping and with timing that I plan to fix - once I get a handle on this timing part of driving the DAC.

Anyone who has any guidance please let me know if this is the right thing to look at ... i'm just looking for a good pointer.

**Comment by marcel-licence** [comment, marked as answer]
If you use portMAX_DELAY the function will block until the complete buffer is written. So it prevents you from overrun but not from underrun. I intentionally use this instead of interrupts etc. to get an audible feedback if the processing in the main loop is too slow and it didn't manage to fill the buffer in time.

**Reply by MichaelPNolan** [reply]
I was going to play with it but I decided to go and make more mundane or fun code for now - i'm still researching optimization and learning various gaps in knowledge like inline ...
